spark/src/main/scala/org/apache/spark/sql/delta/SnapshotManagement.scala

```diff
@@ -628,6 +638,41 @@ trait SnapshotManagement { self: DeltaLog =>
     timestamp => cutoffOpt.forall(timestamp < _)
   }
 
+  /**
+   * Get the newest logSegment, using the previous logSegment as a hint. This is faster than
+   * doing a full update, but it won't work on table overwrites that overwrite the delta log.
+   */
+  def getUpdatedLogSegment(oldLogSegment: LogSegment): (LogSegment, Seq[FileStatus]) = {
+    val newFilesOpt = listDeltaCompactedDeltaAndCheckpointFiles(
+      startVersion = oldLogSegment.version + 1,
+      versionToLoad = None,
+      includeMinorCompactions = spark.conf.get(DeltaSQLConf.DELTALOG_MINOR_COMPACTION_USE_FOR_READS)
+    )
+    val newFiles = newFilesOpt.getOrElse {
+      // An empty listing likely implies a list-after-write inconsistency or that somebody clobbered
+      // the Delta log.
+      return (oldLogSegment, Seq.empty)
+    }
+    val allFiles = (
+      oldLogSegment.checkpointProvider.topLevelFiles ++
+        oldLogSegment.deltas ++
+        newFiles
+      ).toArray
+    val lastCheckpointInfo: Option[LastCheckpointInfo] = None
+    val newLogSegment = getLogSegmentForVersion(
+      versionToLoad = None,
+      files = Some(allFiles),
+      validateLogSegmentWithoutCompactedDeltas = false,
+      lastCheckpointInfo = lastCheckpointInfo,
+      oldCheckpointProviderOpt = Some(oldLogSegment.checkpointProvider)
+    ).getOrElse(oldLogSegment)
+    val fileStatusesOfConflictingCommits = newFiles.filter {
+      case DeltaFile(_, v) if v <= newLogSegment.version => true
+      case _ => false
+    }
```

Comment:
nit
```suggestion
    val fileStatusesOfConflictingCommits = newFiles.collect {
      case DeltaFile(f, v) if v <= newLogSegment.version => f
    }
```

Reply:
Or
```suggestion
    val fileStatusesOfConflictingCommits = newFiles.filter {
      case DeltaFile(_, v) => v <= newLogSegment.version
    }
```

Reply:
The 2nd implementation fails with a MatchError since `newFiles` could have other kind of files also?

Reply:
> The 2nd implementation fails

good point!